### PR TITLE
Add zbot

### PR DESCRIPTION
**Pull request by alik-git**
image

For the runs in the above graph, the number of samples per learning step is 
num_minibatches=64 * num_env_states_per_minibatch=8192 which is 524,288

With stl collisions it took 7.719 hr to get to 100 learning steps
With box collisions it took 1.32 hr to get to 100 learning steps

This means stl collisions gets 1887 samples per second
And box collisions gets 11036 samples per second

Meaning box collisions for zbot is ~5.8x faster.

### CODE REVIEW

File: examples/zbot2/zbot2.mjcf [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+  <asset>
+    <texture type="skybox" builtin="gradient" rgb1="0.3 0.5 0.7" rgb2="0 0 0" width="512" height="3072" />
+    <texture type="2d" name="groundplane" builtin="checker" mark="edge" rgb1="0.2 0.3 0.4" rgb2="0.1 0.2 0.3" markrgb="0.8 0.8 0.8" width="300" height="300" />
+    <material name="groundplane" texture="groundplane" texuniform="true" texrepeat="5 5" reflectance="0.2" />
+    <hfield name="hfield" nrow="2" ncol="2" size="10 10 .05 0.1" />
+  </asset>
+
+  <worldbody>
+    <geom name="floor" type="hfield" hfield="hfield" contype="1" conaffinity="0" material="groundplane" friction="1 0.005 0.0001" pos="0 0 -0.1" />
```
Comment (budzianowski) [inline examples/zbot2/zbot2.mjcf]: no smooth plane?

File: examples/zbot2/walking.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+class ZBot2ActorModel(GaussianActionModel):
+    mlp: MLP
+    action_clipping: float = 20.0
+
+    def __call__(self, obs: FrozenDict[str, Array], cmd: FrozenDict[str, Array]) -> Array:
+        lin_vel_cmd_2 = cmd["linear_velocity_command"]
+        ang_vel_cmd_1 = cmd["angular_velocity_command"]
+        joint_pos_j = obs["joint_position_observation_noisy"]
+        joint_vel_j = obs["joint_velocity_observation_noisy"]
+        # imu_acc_3 = obs["imu_acc_sensor_observation"]
```
Comment (budzianowski) [inline examples/zbot2/walking.py]: add or remove

File: examples/zbot2/walking.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+    actuator_type: str = xax.field(value="mit", help="The type of actuator to use.")
+
+
+class ZBot2WalkingTask(PPOTask[ZBot2WalkingConfig]):
+    def get_environment(self) -> MjxEnv:
+        return MjxEnv(
+            self.config,
+            terminations=[
+                # IllegalContactTerminationBuilder(
```
Comment (budzianowski) [inline examples/zbot2/walking.py]: remove if not used

File: examples/zbot2/robot.xml [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+              <body name="3215_BothFlange_14" pos="0.100000 -0.010000 -0.001260">
+                <joint name="right_knee" type="hinge" class="motor" range="-2.792527 1.221731" axis="0 0 1" />
+                <inertial pos="0.050000 0 -0.017888" mass="0.128213" diaginertia="3.6e-05 0.000155 0.000145" />
+                <geom name="3215_BothFlange_14_visual" material="floating_base_link_material" type="mesh" mesh="3215_BothFlange_14.stl" class="visual" />
+                <body name="FOOT_2" pos="0.100000 0 0.001000" quat="-1.2247452309842457e-07 -0.7071068967259713 0.7071066656470838 1.22474483074382e-07">
+                  <joint name="right_ankle" type="hinge" class="motor" range="-1.570796 1.570796" axis="0 0 1" />
+                  <inertial pos="-0.015078 -0.027835 0.018700" mass="0.278140" diaginertia="4.5e-05 0.000205 0.000182" />
+                  <geom name="FOOT_2_collision" type="mesh" mesh="FOOT_2.collision.stl" class="collision" />
+                  <geom name="FOOT_2_visual" material="L_ARM_MIRROR_1_material" type="mesh" mesh="FOOT_2.stl" class="visual" />
```
Comment (budzianowski) [inline examples/zbot2/robot.xml]: What's the time per step? We want to probably add box collision - see https://github.com/kscalelabs/ksim/pull/52/files#diff-df1e3d8385ed9238d8c49c8a238e5015496ad0a0b954b5c0f9800d8b1a1a4c07R145

File: examples/zbot2/robot.xml [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            </body>
+          </body>
+        </body>
+      </body>
+      <site name="floating_base_link_site" />
+      <camera name="front_camera" mode="track" fovy="90.0" quat="4.329780281177467e-17 4.329780281177466e-17 0.7071067811865475 0.7071067811865476" pos="0.0 2.0 0.5" />
+      <camera name="side_camera" mode="track" fovy="90.0" quat="-0.5 -0.4999999999999999 0.5 0.5000000000000001" pos="-2.0 0.0 0.5" />
+    </body>
+  </worldbody>
+
```
Comment (budzianowski) [inline examples/zbot2/robot.xml:178]: Please add here also explicit contacts list with the floor - see https://github.com/kscalelabs/ksim/pull/52/files#diff-df1e3d8385ed9238d8c49c8a238e5015496ad0a0b954b5c0f9800d8b1a1a4c07R222

File: ksim/utils/mujoco.py [hunk 6]
```diff
@@ -240,10 +240,15 @@ def get_collision_info(contact: mjx.Contact, geom1: int, geom2: int) -> tuple[ja
...
     dist = contact.dist[idx] * mask[idx]
-    normal = (dist < 0) * contact.frame[idx, 0, :3]
+    normal = (dist < 0) * jnp.reshape(contact.frame[idx], (-1,))[:3]
     return dist, normal
 
 
 def geoms_colliding(state: mjx.Data, geom1: int, geom2: int) -> jax.Array:
     """Return True if the two geoms are colliding."""
-    return get_collision_info(state.contact, geom1, geom2)[0] < 0
+    return jax.lax.cond(
```
Comment (alik-git) [inline ksim/utils/mujoco.py:250]: this is because the geoms colliding throws an error about taking argmin of an empty list if there are no collisions, so I perform that check here

File: ksim/utils/mujoco.py [hunk 7]
```diff
@@ -240,10 +240,15 @@ def get_collision_info(contact: mjx.Contact, geom1: int, geom2: int) -> tuple[ja
     mask |= (jnp.array([geom2, geom1]) == contact.geom).all(axis=1)
     idx = jnp.where(mask, contact.dist, 1e4).argmin()
     dist = contact.dist[idx] * mask[idx]
-    normal = (dist < 0) * contact.frame[idx, 0, :3]
+    normal = (dist < 0) * jnp.reshape(contact.frame[idx], (-1,))[:3]
```
Comment (alik-git) [inline ksim/utils/mujoco.py:244]: this is so that this function works even with one env, there the middle dim does not exist and is squeezed out

Comment (budzianowski) [reply ksim/utils/mujoco.py:244]: it would be great to leave a comment like that for the reader

**Comment by budzianowski** [comment]
Can you add to the PR description the time per step with original stls and new collisions boxes with the training setup cause they I much bigger than what I reported with kbot?

File: ksim/builders/resets.py [hunk 8]
```diff
@@ -105,19 +105,39 @@ class XYPositionResetBuilder(ResetBuilder[XYPositionReset]):
...
+                    "No heightfield or plane geom found in the model. Mujoco scene missing floor!"
+                )
+
+            floor_idx = plane_indices[0]
+            x, y = data.model.geom_size[floor_idx][:2]
+            z_pos = data.model.geom_pos[floor_idx][2]
+
+            ztop = z_pos + 0.05
+            zbottom = z_pos - 0.05
+            hfield_data = jnp.full((2, 2), z_pos)
```
Comment (budzianowski) [inline ksim/builders/resets.py:140]: We don't want to make it bigger given the reported issues in the end?

File: examples/zbot2/walking.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+            dt=0.001,
+            max_episode_length=20.0,
+            ctrl_dt=0.008,
+            learning_rate=5e-5,
+            save_every_n_seconds=60 * 4,
+            only_save_most_recent=False,
+            reward_scaling_alpha=0.0,
+            obs_norm_alpha=0.0,
+            scale_rewards=True,
+            # ksim-legacy original setup was dt=0.003 and ctrl_dt=0.012 ~ 83.33 hz
```
Comment (budzianowski) [inline examples/zbot2/walking.py]: Nit: remove

File: examples/zbot2/walking.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+if __name__ == "__main__":
+    # python -m examples.zbot2.walking
+    ZBot2WalkingTask.launch(
+        ZBot2WalkingConfig(
+            num_learning_epochs=8,
+            num_env_states_per_minibatch=8192,
+            num_minibatches=64,
+            num_envs=2048,
+            dt=0.001,
+            max_episode_length=20.0,
```
Comment (budzianowski) [inline examples/zbot2/walking.py]: I don't think it's used anymore

**Review by budzianowski — APPROVED** [review]
Thanks for addressing all the comments - lgtm. Can we add the step timing with stl and box collisions though?

File: ksim/builders/resets.py (hunk 8 above)
Comment (alik-git) [reply ksim/builders/resets.py:140]: sorry this code is actually outdated and I was going to update it in my next PR for the reset functions. For now I am just returning "pretend" hfield data for the plane based floor. This "pretend" hfield data is only used for sampling the height of the reset point.
